openhands_aci/navigator/navigator.py

```diff
@@ -0,0 +1,297 @@
+        ident2refrels = defaultdict(
+            list
+        )  # symbol identifier -> list of its references' relative file paths
+        identwrel2deftags = defaultdict(
+            set
+        )  # (relative file, symbol identifier) -> set of its DEF tags
+        identwrel2reftags = defaultdict(
+            set
+        )  # (relative file, symbol identifier) -> set of its REF tags
+
+        all_abs_files_iter = (
+            tqdm(all_abs_files, desc='Parsing tags', unit='file')
+            if self.show_progress
+            else all_abs_files
+        )
+        for abs_file in all_abs_files_iter:
+            rel_file = self.path_utils.get_relative_path_str(abs_file)
+            parsed_tags = self.ts_parser.get_tags_from_file(abs_file, rel_file)
+
+            for parsed_tag in parsed_tags:
+                if parsed_tag.tag_kind == TagKind.DEF:
+                    ident2defrels[parsed_tag.node_content].add(rel_file)
+                    identwrel2deftags[(rel_file, parsed_tag.node_content)].add(
+                        parsed_tag
+                    )
+                if parsed_tag.tag_kind == TagKind.REF:
+                    ident2refrels[parsed_tag.node_content].append(rel_file)
+                    identwrel2reftags[(rel_file, parsed_tag.node_content)].add(
+                        parsed_tag
+                    )
+
+        return ident2defrels, ident2refrels, identwrel2deftags, identwrel2reftags
+
+    def _tag_list_to_tree(self, tags: list[ParsedTag], use_end_line=False) -> str:
+        """
+        Convert a list of ParsedTag objects to a tree str representation.
+        """
+        if not tags:
+            return ''
+
+        cur_rel_file, cur_abs_file = '', ''
+        lines_of_interest: list[int] = []
+        output = ''
+
+        dummy_tag = ParsedTag(
+            abs_path='',
+            rel_path='',
+            node_content='',
+            tag_kind=TagKind.DEF,
+            start_line=0,
+            end_line=0,
+        )
+        for tag in tags + [dummy_tag]:  # Add dummy tag to trigger last file output
+            if tag.rel_path != cur_rel_file:
+                if lines_of_interest:
+                    output += cur_rel_file + ':\n'
+                    output += self._render_tree(
+                        cur_abs_file, cur_rel_file, lines_of_interest
+                    )
+                    lines_of_interest = []
+                elif cur_rel_file:  # No line of interest
+                    output += '\n' + cur_rel_file + ':\n'
+
+                cur_abs_file = tag.abs_path
+                cur_rel_file = tag.rel_path
+
+            lines_of_interest += (
+                list(range(tag.start_line, tag.end_line + 1))
+                if use_end_line
+                else [tag.start_line]
+            )
+
+        # Truncate long lines in case we get minified js or something else crazy
+        output = '\n'.join(line[:150] for line in output.splitlines())
```

Comment:
can you get start and end positions from the parsed tags to help with this sort of thing (minified js or what have you)?